src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,974 @@
+
+        /// <summary>
+        /// Determines if the expression is something that should have code actions displayed for it.
+        /// Depends upon the identifiers in the expression mapping back to parameters.
+        /// Does not handle params parameters.
+        /// </summary>
+        private static async Task<(bool shouldDisplay, bool containsClassExpression)> ShouldExpressionDisplayCodeActionAsync(Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var containsClassSpecificStatement = false;
+            var syntaxKinds = document.GetRequiredLanguageService<ISyntaxKindsService>();
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IRangeVariableSymbol or ILocalSymbol)
+                {
+                    return (false, false);
+                }
+
+                if (symbol is IParameterSymbol parameter)
+                {
+                    if (parameter.IsParams)
+                    {
+                        return (false, false);
+                    }
+                }
+
+                if (variable.RawKind == syntaxKinds.ThisExpression || variable.RawKind == syntaxKinds.BaseExpression)
+                {
+                    containsClassSpecificStatement = true;
+                }
+            }
+
+            return (true, containsClassSpecificStatement);
+        }
+
+        /// <summary>
+        /// Creates new code actions for each introduce parameter possibility.
+        /// Does not create actions for overloads/trampoline if there are optional parameters or if the methodSymbol
+        /// is a constructor.
+        /// </summary>
+        private async Task<(ImmutableArray<CodeAction> actions, ImmutableArray<CodeAction> actionsAllOccurrences)?> GetActionsAsync(Document document,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            CancellationToken cancellationToken)
+        {
+            var (shouldDisplay, containsClassExpression) = await ShouldExpressionDisplayCodeActionAsync(
+                document, expression, cancellationToken).ConfigureAwait(false);
+            if (!shouldDisplay)
+            {
+                return null;
+            }
+
+            using var actionsBuilder = TemporaryArray<CodeAction>.Empty;
+            using var actionsBuilderAllOccurrences = TemporaryArray<CodeAction>.Empty;
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var singleLineExpression = syntaxFacts.ConvertToSingleLine(expression);
+            var nodeString = singleLineExpression.ToString();
```

Comment:
this seems weird.  you are doing this in the toplevel code actions as well.  this feels like a lot of redundancy.  can you show me a picture of what the lightbulb items look like now?

Reply:
![image](https://user-images.githubusercontent.com/40616383/115071273-e8112a00-9eaa-11eb-8765-ad5d54202965.png)
but imagine "callsites" is "call sites"

Reply:
"into new overload of x+y"?

Reply:
changed to into new overload